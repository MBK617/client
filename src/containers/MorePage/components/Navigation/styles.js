import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  item: {
    padding: 10,
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderBottomWidth: 3,
    borderColor: 'rgba(0, 0, 0, 0.1)'
  },
  itemName: {
    fontSize: 18,
    marginLeft: 15,
  },
  itemInfo: {
    flexDirection: 'row',
    alignItems: 'center'
  }
});

export default styles;