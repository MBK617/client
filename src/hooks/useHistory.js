import { useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { createSelector } from 'reselect';
import { popHistory, pushHistory } from 'containers/App/actions';

const selectApp = createSelector(
  state => state.app,
  substate => substate.toJS()
);

const useHistory = () => {
  const { history } = useSelector(selectApp);
  const dispatch = useDispatch();

  const path = useMemo(() => {
    return history.length > 0 ? history[0] : '';
  }, [history]);

  const goTo = (value) => {
    let newPath = value;
    if(newPath.substr(0, 2) === './') {
      newPath = `${path}/${newPath}`
    }
    if(path !== newPath) {
      dispatch(pushHistory(newPath));
    }
  }

  const goBack = () => {
    if(history.length > 1) {
      dispatch(popHistory());
    }
  }

  return {
    history,
    path,
    goTo,
    goBack
  };
}

export default useHistory;