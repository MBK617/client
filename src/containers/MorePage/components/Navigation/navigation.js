import React from 'react';
import { 
  View, 
  Text, 
  TouchableHighlight, 
  Linking, 
  Alert 
} from "react-native";
import { 
  MaterialCommunityIcons as Icon, 
  Feather as FeatherIcon 
} from '@expo/vector-icons'

import useHistory from 'hooks/useHistory';

import styles from './styles';

const Navigation = ({ routes }) => {
  const { goTo } = useHistory();

  return (
    <View style={styles.root}>
      {routes.map(route =>  
        route.show && 
          <TouchableHighlight 
            underlayColor="rgba(0,0,0,0.1)" 
            key={route.name} 
            onPress={() => {
              if(route.key) goTo(`more/${route.key}`);
              if(route.external) Linking.canOpenURL(route.external).then(supported => {
                if(supported) {
                  Linking.openURL(route.external);
                } else {
                  Alert.alert("Can't open this link in your default browser.")
                }
              })
            }}
          >
            <View style={styles.item} >
              <View style={styles.itemInfo}>
                <Icon 
                  name={route.icon} 
                  size={40} 
                />
                <Text
                  style={styles.itemName}
                >
                  {route.name}
                </Text>
              </View>
              {route.external ? <FeatherIcon
                name= "external-link"
                size={32}
                style={{ marginLeft: -10 }}
              /> : <Icon
                name="chevron-right"
                size={42}
              /> }
            </View>
          </TouchableHighlight>
      )}
    </View>
  );
}

export default Navigation;