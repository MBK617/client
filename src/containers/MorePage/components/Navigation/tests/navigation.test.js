import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react-native';
import useHistory from 'hooks/useHistory';

import Navigation from '..';

jest.mock('@expo/vector-icons');
let mockOpenURL = jest.fn();

jest.mock('react-native/Libraries/Linking/Linking', () => ({
  canOpenURL: jest.fn(() => Promise.resolve(true)),
  openURL: mockOpenURL,
}));

jest.mock('hooks/useHistory');
jest.mock('hooks/useLoginStatus', () => {
  return jest.fn(() => true);
});

describe('<Navigation />', () => {
  let rendered, goTo;
  const testRoutes = [
    {
      name: "Example",
      path: 'example',
      icon: 'account-circle',
      show: true
    },
    {
      name: "External",
      icon: 'heart',
      external: 'https://google.com',
      show: true
    }
  ]

  beforeEach(() => {
    goTo = jest.fn();
    useHistory.mockImplementation(() => ({
      goTo
    }));

    rendered = render(<Navigation routes={testRoutes}/>);
  })

  afterEach(cleanup);

  it('should render correctly', () => {
    const tree = rendered.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render navigation buttons', () => {
    testRoutes.forEach((route) => {
      expect(rendered.queryByText(route.name)).toBeTruthy();
    })
  });

  it('should not render navigation buttons for routes with !show', () => {
    rendered = render(
      <Navigation 
        routes={[{
          name: "Example",
          path: 'example',
          icon: 'account-circle',
          show: false
        }]}
      />
    );
    expect(rendered.queryByText("Example")).toBeFalsy();
  });

  it('should navigate on button clicks', () => {
    const button = rendered.queryByText("Example");
    fireEvent.press(button);
    expect(goTo).toHaveBeenCalledWith('more/example');
  });

  it('should open link on button click', (done) => {
    const button = rendered.queryByText('External');
    fireEvent.press(button);
    setTimeout(() => {
      expect(mockOpenURL).toHaveBeenCalledWith('https://google.com');
      done();
    }, 100);
  });

});