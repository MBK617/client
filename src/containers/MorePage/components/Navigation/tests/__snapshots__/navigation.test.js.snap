// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> should render correctly 1`] = `
<View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 3,
          "borderColor": "rgba(0, 0, 0, 0.1)",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginLeft": 15,
            }
          }
        >
          Example
        </Text>
      </View>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 3,
          "borderColor": "rgba(0, 0, 0, 0.1)",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginLeft": 15,
            }
          }
        >
          External
        </Text>
      </View>
    </View>
  </View>
</View>
`;
